lib/timers.js

```diff
@@ -477,47 +477,109 @@ exports.clearImmediate = function(immediate) {
 
 var unrefList, unrefTimer;
 
+function _makeTimerTimeout(timer) {
+  var domain = timer.domain;
+  var msecs = timer._idleTimeout;
+
+  // Timer has been unenrolled by another timer that fired at the same time,
+  // so don't make it timeout.
+  if (!msecs || msecs < 0)
+    return;
+
+  if (!timer._onTimeout)
+    return;
+
+  if (domain) {
+    if (domain._disposed)
+      return;
+
+    domain.enter();
+  }
+
+  debug('unreftimer firing timeout');
+  L.remove(timer);
+  timer._called = true;
+  _runOnTimeout(timer);
+
+  if (domain)
+    domain.exit();
+}
+
+function _runOnTimeout(timer) {
+  try {
+    var threw = true;
+    timer._onTimeout();
+    threw = false;
+  } finally {
+    if (threw) process.nextTick(unrefTimeout);
```

Comment:
Correct me if I'm wrong, but isn't this also what regular timers do?

https://github.com/nodejs/node/blob/master/lib/timers.js#L93-L103

I tried to "fix" that to make it more understandable at one point but then it didn't really work..